// @clerk/nextjs/dist/cjs/app-beta/client/ui-components.js
"use strict";
"use client";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ui_components_exports = {};
__export(ui_components_exports, {
  CreateOrganization: () => import_clerk_react.CreateOrganization,
  OrganizationProfile: () => import_clerk_react.OrganizationProfile,
  OrganizationSwitcher: () => import_clerk_react.OrganizationSwitcher,
  SignIn: () => import_clerk_react.SignIn,
  SignUp: () => import_clerk_react.SignUp,
  UserButton: () => import_clerk_react.UserButton,
  UserProfile: () => import_clerk_react.UserProfile
});
module.exports = __toCommonJS(ui_components_exports);
var import_clerk_react = require("@clerk/clerk-react");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CreateOrganization,
  OrganizationProfile,
  OrganizationSwitcher,
  SignIn,
  SignUp,
  UserButton,
  UserProfile
});
