// @clerk/clerk-react/dist/cjs/components/uiComponents.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var uiComponents_exports = {};
__export(uiComponents_exports, {
  CreateOrganization: () => CreateOrganization,
  OrganizationList: () => OrganizationList,
  OrganizationProfile: () => OrganizationProfile,
  OrganizationSwitcher: () => OrganizationSwitcher,
  SignIn: () => SignIn,
  SignUp: () => SignUp,
  UserButton: () => UserButton,
  UserProfile: () => UserProfile
});
module.exports = __toCommonJS(uiComponents_exports);
var import_react = __toESM(require("react"));
var import_withClerk = require("./withClerk");
class Portal extends import_react.default.PureComponent {
  constructor() {
    super(...arguments);
    this.portalRef = import_react.default.createRef();
  }
  componentDidUpdate(prevProps) {
    if (prevProps.props.appearance !== this.props.props.appearance) {
      this.props.updateProps({ node: this.portalRef.current, props: this.props.props });
    }
  }
  componentDidMount() {
    if (this.portalRef.current) {
      this.props.mount(this.portalRef.current, this.props.props);
    }
  }
  componentWillUnmount() {
    if (this.portalRef.current) {
      this.props.unmount(this.portalRef.current);
    }
  }
  render() {
    return /* @__PURE__ */ import_react.default.createElement("div", { ref: this.portalRef });
  }
}
const SignIn = (0, import_withClerk.withClerk)(({ clerk, ...props }) => {
  return /* @__PURE__ */ import_react.default.createElement(
    Portal,
    {
      mount: clerk.mountSignIn,
      unmount: clerk.unmountSignIn,
      updateProps: clerk.__unstable__updateProps,
      props
    }
  );
}, "SignIn");
const SignUp = (0, import_withClerk.withClerk)(({ clerk, ...props }) => {
  return /* @__PURE__ */ import_react.default.createElement(
    Portal,
    {
      mount: clerk.mountSignUp,
      unmount: clerk.unmountSignUp,
      updateProps: clerk.__unstable__updateProps,
      props
    }
  );
}, "SignUp");
const UserProfile = (0, import_withClerk.withClerk)(({ clerk, ...props }) => {
  return /* @__PURE__ */ import_react.default.createElement(
    Portal,
    {
      mount: clerk.mountUserProfile,
      unmount: clerk.unmountUserProfile,
      updateProps: clerk.__unstable__updateProps,
      props
    }
  );
}, "UserProfile");
const UserButton = (0, import_withClerk.withClerk)(({ clerk, ...props }) => {
  return /* @__PURE__ */ import_react.default.createElement(
    Portal,
    {
      mount: clerk.mountUserButton,
      unmount: clerk.unmountUserButton,
      updateProps: clerk.__unstable__updateProps,
      props
    }
  );
}, "UserButton");
const OrganizationProfile = (0, import_withClerk.withClerk)(({ clerk, ...props }) => {
  return /* @__PURE__ */ import_react.default.createElement(
    Portal,
    {
      mount: clerk.mountOrganizationProfile,
      unmount: clerk.unmountOrganizationProfile,
      updateProps: clerk.__unstable__updateProps,
      props
    }
  );
}, "OrganizationProfile");
const CreateOrganization = (0, import_withClerk.withClerk)(({ clerk, ...props }) => {
  return /* @__PURE__ */ import_react.default.createElement(
    Portal,
    {
      mount: clerk.mountCreateOrganization,
      unmount: clerk.unmountCreateOrganization,
      updateProps: clerk.__unstable__updateProps,
      props
    }
  );
}, "CreateOrganization");
const OrganizationSwitcher = (0, import_withClerk.withClerk)(({ clerk, ...props }) => {
  return /* @__PURE__ */ import_react.default.createElement(
    Portal,
    {
      mount: clerk.mountOrganizationSwitcher,
      unmount: clerk.unmountOrganizationSwitcher,
      updateProps: clerk.__unstable__updateProps,
      props
    }
  );
}, "OrganizationSwitcher");
const OrganizationList = (0, import_withClerk.withClerk)(({ clerk, ...props }) => {
  return /* @__PURE__ */ import_react.default.createElement(
    Portal,
    {
      mount: clerk.mountOrganizationList,
      unmount: clerk.unmountOrganizationList,
      updateProps: clerk.__unstable__updateProps,
      props
    }
  );
}, "OrganizationList");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CreateOrganization,
  OrganizationList,
  OrganizationProfile,
  OrganizationSwitcher,
  SignIn,
  SignUp,
  UserButton,
  UserProfile
});
