// @clerk/shared/dist/cjs/hooks/contexts.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var contexts_exports = {};
__export(contexts_exports, {
  ClerkInstanceContext: () => ClerkInstanceContext,
  ClientContext: () => ClientContext,
  OrganizationContext: () => OrganizationContext,
  OrganizationProvider: () => OrganizationProvider,
  SessionContext: () => SessionContext,
  UserContext: () => UserContext,
  useClerkInstanceContext: () => useClerkInstanceContext,
  useClientContext: () => useClientContext,
  useOrganizationContext: () => useOrganizationContext,
  useSessionContext: () => useSessionContext,
  useUserContext: () => useUserContext
});
module.exports = __toCommonJS(contexts_exports);
var import_react = __toESM(require("react"));
var import_clerk_swr = require("./clerk-swr");
var import_clerk_swr2 = require("./clerk-swr");
var import_createContextAndHook = require("./createContextAndHook");
(0, import_clerk_swr.disableSWRDevtools)();
const [ClerkInstanceContext, useClerkInstanceContext] = (0, import_createContextAndHook.createContextAndHook)("ClerkInstanceContext");
const [UserContext, useUserContext] = (0, import_createContextAndHook.createContextAndHook)("UserContext");
const [ClientContext, useClientContext] = (0, import_createContextAndHook.createContextAndHook)("ClientContext");
const [SessionContext, useSessionContext] = (0, import_createContextAndHook.createContextAndHook)(
  "SessionContext"
);
const [OrganizationContextInternal, useOrganizationContext] = (0, import_createContextAndHook.createContextAndHook)("OrganizationContext");
const OrganizationProvider = ({
  children,
  organization,
  lastOrganizationMember,
  lastOrganizationInvitation,
  swrConfig
}) => {
  return /* @__PURE__ */ import_react.default.createElement(import_clerk_swr2.SWRConfig, { value: swrConfig }, /* @__PURE__ */ import_react.default.createElement(
    OrganizationContextInternal.Provider,
    {
      value: {
        value: {
          organization,
          lastOrganizationMember,
          lastOrganizationInvitation
        }
      }
    },
    children
  ));
};
const OrganizationContext = OrganizationProvider;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ClerkInstanceContext,
  ClientContext,
  OrganizationContext,
  OrganizationProvider,
  SessionContext,
  UserContext,
  useClerkInstanceContext,
  useClientContext,
  useOrganizationContext,
  useSessionContext,
  useUserContext
});
