// @azure/openai/dist-esm/src/api/deserializers.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
function getPromptFilterResult(body) {
    var _a;
    const res = (_a = body["prompt_annotations"]) !== null && _a !== void 0 ? _a : body["prompt_filter_results"];
    return !res
        ? {}
        : {
            promptFilterResults: res.map((p) => (Object.assign({ promptIndex: p["prompt_index"] }, (!p.content_filter_results
                ? {}
                : {
                    contentFilterResults: deserializeContentFilter(p.content_filter_results),
                })))),
        };
}
export function getCompletionsResult(body) {
    var _a;
    return Object.assign(Object.assign({ id: body["id"], created: new Date(body["created"]) }, getPromptFilterResult(body)), { choices: ((_a = body["choices"]) !== null && _a !== void 0 ? _a : []).map((p) => (Object.assign(Object.assign({ text: p["text"], index: p["index"] }, (!p.content_filter_results
            ? {}
            : {
                contentFilterResults: deserializeContentFilter(p.content_filter_results),
            })), { logprobs: p.logprobs === null
                ? null
                : {
                    tokens: p.logprobs["tokens"],
                    tokenLogprobs: p.logprobs["token_logprobs"],
                    topLogprobs: p.logprobs["top_logprobs"],
                    textOffset: p.logprobs["text_offset"],
                }, finishReason: p["finish_reason"] }))) });
}
export function getChatCompletionsResult(body) {
    var _a;
    return Object.assign(Object.assign({ id: body["id"], created: new Date(body["created"]), choices: ((_a = body["choices"]) !== null && _a !== void 0 ? _a : []).map((p) => (Object.assign(Object.assign(Object.assign(Object.assign({}, (!p.message ? {} : { message: _deserializeMessage(p.message) })), { index: p["index"], finishReason: p["finish_reason"] }), (!p.delta ? {} : { delta: _deserializeMessage(p.delta) })), (!p.content_filter_results
            ? {}
            : { contentFilterResults: deserializeContentFilter(p.content_filter_results) })))) }, getPromptFilterResult(body)), (!body["usage"]
        ? {}
        : {
            usage: {
                completionTokens: body["usage"].completion_tokens,
                promptTokens: body["usage"].prompt_tokens,
                totalTokens: body["usage"].total_tokens,
            },
        }));
}
function _deserializeMessage(message) {
    var _a, _b;
    return Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({}, (!message["role"] ? {} : { role: message["role"] })), (!message["content"] ? {} : { content: message["content"] })), (!message["name"] ? {} : { name: message["name"] })), (!message.function_call
        ? {}
        : {
            functionCall: {
                name: (_a = message.function_call) === null || _a === void 0 ? void 0 : _a["name"],
                arguments: (_b = message.function_call) === null || _b === void 0 ? void 0 : _b["arguments"],
            },
        })), (!message.context
        ? {}
        : {
            context: Object.assign({}, (!message.context.messages
                ? {}
                : {
                    messages: message.context.messages.map((m) => {
                        return _deserializeMessage(m);
                    }),
                })),
        }));
}
function deserializeContentFilter(result) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j;
    if (result.error) {
        return {
            error: {
                code: result.error.code,
                message: result.error.message,
                details: (_a = result.error.details) !== null && _a !== void 0 ? _a : [],
            },
        };
    }
    return Object.assign(Object.assign(Object.assign(Object.assign({}, (!result.sexual
        ? {}
        : {
            sexual: {
                severity: (_b = result.sexual) === null || _b === void 0 ? void 0 : _b["severity"],
                filtered: (_c = result.sexual) === null || _c === void 0 ? void 0 : _c["filtered"],
            },
        })), (!result.violence
        ? {}
        : {
            violence: {
                severity: (_d = result.violence) === null || _d === void 0 ? void 0 : _d["severity"],
                filtered: (_e = result.violence) === null || _e === void 0 ? void 0 : _e["filtered"],
            },
        })), (!result.hate
        ? {}
        : {
            hate: {
                severity: (_f = result.hate) === null || _f === void 0 ? void 0 : _f["severity"],
                filtered: (_g = result.hate) === null || _g === void 0 ? void 0 : _g["filtered"],
            },
        })), (!result.self_harm
        ? {}
        : {
            selfHarm: {
                severity: (_h = result.self_harm) === null || _h === void 0 ? void 0 : _h["severity"],
                filtered: (_j = result.self_harm) === null || _j === void 0 ? void 0 : _j["filtered"],
            },
        }));
}
